File: benches/three/mandelbrot_f64.c
/* The same picture in double precision -- the row lex-sys cannot fill.
 *
 * Not a fair race against the fixed-point versions and not meant to be.
 * It is here to answer a different question: what does the *absence* of
 * a float type cost, given that a program wanting this picture would
 * write this and not Q16.16. See `docs/against-c-and-rust.md` §4. */
#include <stdio.h>

__attribute__((noinline))
static long escape(double cx, double cy, long maxiter) {
    double zx = 0, zy = 0, zx2 = 0, zy2 = 0;
    long i = 0;
    while (i < maxiter && zx2 + zy2 <= 4.0) {
        double zxy = zx * zy;
        zx = zx2 - zy2 + cx;
        zy = 2 * zxy + cy;
        zx2 = zx * zx;
        zy2 = zy * zy;
        i++;
    }
    return i;
}

__attribute__((noinline))
static long grid(long width, long height, long maxiter) {
    long total = 0;
    for (long py = 0; py < height; py++) {
        double cy = ((double)py * 2.5 / (double)height) - 1.25;
        for (long px = 0; px < width; px++) {
            double cx = ((double)px * 2.5 / (double)width) - 2.0;
            total += escape(cx, cy, maxiter);
        }
    }
    return total;
}

int main(void) {
    printf("%ld\n", grid(400, 400, 1000));
    return 0;
}

File: benches/three/mandelbrot_f64.rs
// The same picture in double precision -- the row lex-sys cannot fill.
// `docs/against-c-and-rust.md` §4.

#[inline(never)]
fn escape(cx: f64, cy: f64, maxiter: i64) -> i64 {
    let (mut zx, mut zy, mut zx2, mut zy2) = (0.0f64, 0.0f64, 0.0f64, 0.0f64);
    let mut i = 0i64;
    while i < maxiter && zx2 + zy2 <= 4.0 {
        let zxy = zx * zy;
        zx = zx2 - zy2 + cx;
        zy = 2.0 * zxy + cy;
        zx2 = zx * zx;
        zy2 = zy * zy;
        i += 1;
    }
    i
}

#[inline(never)]
fn grid(width: i64, height: i64, maxiter: i64) -> i64 {
    let mut total = 0i64;
    for py in 0..height {
        let cy = (py as f64) * 2.5 / (height as f64) - 1.25;
        for px in 0..width {
            let cx = (px as f64) * 2.5 / (width as f64) - 2.0;
            total += escape(cx, cy, maxiter);
        }
    }
    total
}

fn main() {
    println!("{}", grid(400, 400, 1000));
}
